vms/platformvm/validators/manager.go

```diff
@@ -198,146 +225,162 @@ func (m *manager) getValidatorSetFrom(
 }
 
 func (m *manager) makeSubnetValidatorSet(
-	currentHeight uint64,
+	ctx context.Context,
 	targetHeight uint64,
 	subnetID ids.ID,
-) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
-	currentValidators, ok := m.cfg.Validators.Get(subnetID)
-	if !ok {
-		currentValidators = validators.NewSet()
-		if err := m.state.ValidatorSet(subnetID, currentValidators); err != nil {
-			return nil, err
-		}
+) (map[ids.NodeID]*validators.GetValidatorOutput, uint64, error) {
+	subnetValidatorSet, primaryValidatorSet, currentHeight, err := m.getCurrentValidatorSets(ctx, subnetID)
+	if err != nil {
+		return nil, 0, err
 	}
-
-	// Node ID --> Validator information for the node validating the Subnet.
-	subnetValidatorSet := currentValidators.Map()
-
-	// Rebuild subnet validators at [targetHeight]
-	for diffHeight := currentHeight; diffHeight > targetHeight; diffHeight-- {
-		weightDiffs, err := m.state.GetValidatorWeightDiffs(diffHeight, subnetID)
-		if err != nil {
-			return nil, err
-		}
-
-		for nodeID, weightDiff := range weightDiffs {
-			if err := applyWeightDiff(subnetValidatorSet, nodeID, weightDiff); err != nil {
-				return nil, err
-			}
-		}
+	if currentHeight < targetHeight {
+		return nil, 0, database.ErrNotFound
 	}
 
-	// Get the public keys for all the validators at [targetHeight]
-	primarySet, err := m.getValidatorSetFrom(currentHeight, targetHeight, constants.PrimaryNetworkID)
+	// Rebuild subnet validators at [targetHeight]
+	//
+	// Note: Height h contains the diffs from height h-1 to h. So, to rebuild
+	// validators state at height h, we apply diffs till height h+1.
+	lastDiffHeight := targetHeight + 1
+	err = m.state.ApplyValidatorWeightDiffs(
+		ctx,
+		subnetValidatorSet,
+		currentHeight,
+		lastDiffHeight,
+		subnetID,
+	)
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
 
-	// Update the subnet validator set to include the public keys at
-	// [targetHeight].
-	for nodeID, subnetValidator := range subnetValidatorSet {
-		primaryValidator, ok := primarySet[nodeID]
+	for nodeID, vdr := range subnetValidatorSet {
+		primaryVdr, ok := primaryValidatorSet[nodeID]
 		if !ok {
 			// This should never happen
 			m.log.Error(ErrMissingValidator.Error(),
-				zap.Stringer("nodeID", nodeID),
 				zap.Stringer("subnetID", subnetID),
+				zap.Stringer("nodeID", vdr.NodeID),
 			)
-			return nil, ErrMissingValidator
+			return nil, 0, ErrMissingValidator
 		}
-		subnetValidator.PublicKey = primaryValidator.PublicKey
+
+		vdr.PublicKey = primaryVdr.PublicKey
 	}
 
-	return subnetValidatorSet, nil
+	err = m.state.ApplyValidatorPublicKeyDiffs(
+		ctx,
+		subnetValidatorSet,
+		currentHeight,
+		lastDiffHeight,
+	)
+	return subnetValidatorSet, currentHeight, err
 }
 
-func applyWeightDiff(
-	targetSet map[ids.NodeID]*validators.GetValidatorOutput,
-	nodeID ids.NodeID,
-	weightDiff *state.ValidatorWeightDiff,
-) error {
-	vdr, ok := targetSet[nodeID]
+func (m *manager) getCurrentValidatorSets(
```

Comment:
This is separated into its own function because this function will need a P-chain lock held to get a consistent view of the validator set + height.